### ISSUE TITLE
Handle caching is incompatible with shared memory routing

### ISSUE THREAD

**Issue by carns** [mochi-hpc/mochi-margo#47]
In GitLab by @marcvef on Dec 7, 2018, 12:14

Note: This issue only applies to situations where automatic shared memory routing is used.

tl;dr: Margo handle cache is initialized with handles of the user-defined RPC protocol. When `margo_create()` is called with an na+sm address, it causes `margo_handle_cache_get()` to always fail because `HG_Reset()` expects an na+sm handle which margo's handle cache never provides. Therefore, local communication never benefits from margo's handle cache. 

For the explanation below we exemplarily use ofi+tcp as the main RPC protocol which was used to initialize the margo client. In addition, `auto_sm` is enabled. The margo server is started on the same machine and accepts ofi+tcp connections as well as implicitly na+sm connections because `auto_sm` is enabled as well. Since we only use one node in this example na+sm is always used automatically.

Each time `margo_create()` is called, margo acquires a handle which has been initialized in `margo_handle_cache_init()` with ofi+tcp from the handle cache. This handle is then used with `HG_Reset()` and with an address which was automatically set to na+sm by `margo_addr_lookup()` earlier due to automatic shared memory routing in mercury. Therefore, `HG_Reset()` is called with an ofi+tcp handle and an na+sm address. Because `HG_Reset()` can only reuse handles when its address type doesn't change, it causes the following mercury error as the handle cannot be reset with the given address:

```
# HG -- Error -- /home/evie/adafs/git/mercury/src/mercury_core.c:4570
 # HG_Core_reset(): Cannot reset handle to a different address NA class
# HG -- Error -- /home/evie/adafs/git/mercury/src/mercury.c:1966
 # HG_Reset(): Could not reset core HG handle
```
Then margo proceeds with putting the same handle back into the margo cache and calling `HG_Create()` to manually create a shared memory handle. Once this handle will be destroyed in `margo_destroy()`, it will also be discarded and not used in the cache because it was manually created.

This cycle then repeats forever and no handle in margo's cache can ever be used for na+sm communication, accompanied with above error message in each call.

A possible solution would be to use two handle caches if `auto_sm` is enabled and then use the correct cache based on the incoming address.

**Comment by carns** [comment]
In GitLab by @marcvef on Dec 7, 2018, 12:18

changed the description

**Comment by carns** [comment]
In GitLab by @carns on Dec 14, 2018, 10:30

@soumagne do you have a suggestion for how to handle this?  We have a function in margo that looks like this:

```
hg_return_t margo_create(margo_instance_id mid, hg_addr_t addr,
    hg_id_t id, hg_handle_t *handle)
```

That function is analogous to HG_Create(), but internally it draws from a pool of preallocated handles (if available) which gives us a performance boost if we are frequently creating and destroying handles.  We just reset the handles each time we reuse them.  @marcvef has correctly pointed out that this fails if the handles actually go to different transports under the covers, and in the auto_sm case we can't inspect the addr string either.

We could work around it (and retain the cache) if there were a programmatic way to tell what transport an addr/handle was going to use.  Or maybe even if we just got a special return code from HG_Reset() to tell us to handle this differently when it happens?

**Comment by carns** [comment]
In GitLab by @soumagne on Dec 14, 2018, 12:37

Yes I see... I could definitely return a specific error code that could tell you that or you could also overload the HG handle to attach/cache that information using HG_Class_set_handle_create_callback() and HG_Set/get_data().

**Comment by carns** [comment]
In GitLab by @carns on Jan 7, 2019, 13:57

Just a quick update: @soumagne and I talked about this some more this morning. We are going to look into if it's feasible to make HG_Reset() work correctly in this case.  That makes the most sense intuitively (since otherwise the choice of primar/shm routing is transparent), but we have to see what the technical challenges are.  If we can get that to work then the Margo level handle cache doesn't necessarily need to change.

**Comment by carns** [comment]
In GitLab by @marcvef on Jan 8, 2019, 08:39

Thanks for the update, @carns ! I agree, this sounds like a good solution. If we can help with testing, let us know. We can immediately verify if it works when you have a working solution.

**Comment by carns** [comment]
In GitLab by @soumagne on Feb 11, 2019, 18:21

This should now be fixed as part of https://github.com/mercury-hpc/mercury/issues/272

**Comment by carns** [comment]
In GitLab by @carns on Feb 12, 2019, 10:35

Excellent, thanks @soumagne 

Margo could still consider an optimization to separate handles based on transport (it will always be cheaper to reset a handle on its native transport than switch back and forth).  I don't have a good feel for how much performance would be gained by that yet, though.

**Comment by carns** [comment]
In GitLab by @carns on Feb 12, 2019, 10:35

closed

**Comment by carns** [comment]
In GitLab by @soumagne on Feb 12, 2019, 12:10

Just a comment on that, in that case not the entire handle will be re-allocated, only the attached NA resources (in/out buffers, op IDs).
